sycl/source/detail/device_filter.cpp

```diff
@@ -0,0 +1,134 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/sycl/detail/device_filter.hpp>
+#include <CL/sycl/info/info_desc.hpp>
+#include <detail/config.hpp>
+#include <detail/device_impl.hpp>
+
+#include <cstring>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+device_filter::device_filter(const std::string &FilterString) {
+  const std::array<std::pair<std::string, info::device_type>, 5>
+      SyclDeviceTypeMap = {{{"host", info::device_type::host},
+                            {"cpu", info::device_type::cpu},
+                            {"gpu", info::device_type::gpu},
+                            {"acc", info::device_type::accelerator},
+                            {"*", info::device_type::all}}};
+  const std::array<std::pair<std::string, backend>, 4> SyclBeMap = {
+      {{"opencl", backend::opencl},
+       {"level_zero", backend::level_zero},
+       {"cuda", backend::cuda},
+       {"*", backend::all}}};
+
+  // handle the optional 1st field of the filter, backend
+  size_t Cursor = 0;
+  size_t ColonPos = FilterString.find(":", Cursor);
+  // check if the first entry matches with a known backend type
+  auto It = std::find_if(
+      std::begin(SyclBeMap), std::end(SyclBeMap),
+      [=, &Cursor](const std::pair<std::string, backend> &Element) {
+        size_t Found = FilterString.find(Element.first, Cursor);
+        if (Found != std::string::npos) {
+          Cursor = Found;
+          return true;
+        }
+        return false;
+      });
+  // if no match is found, set the backend type backend::all
+  // which actually means 'any backend' will be a match.
+  if (It == SyclBeMap.end()) {
+    Backend = backend::all;
+  } else {
+    Backend = It->second;
+    if (ColonPos != std::string::npos) {
+      Cursor = ColonPos + 1;
+    } else {
+      Cursor = Cursor + It->first.size();
+    }
+  }
+
+  // handle the optional 2nd field of the filter, device type
+  // check if the 2nd entry matches with any known device type.
+  auto Iter = std::find_if(
+      std::begin(SyclDeviceTypeMap), std::end(SyclDeviceTypeMap),
+      [=, &Cursor](const std::pair<std::string, info::device_type> &Element) {
+        size_t Found = FilterString.find(Element.first, Cursor);
+        if (Found != std::string::npos) {
+          Cursor = Found;
+          return true;
+        }
+        return false;
+      });
+  // if no match is found, set device_type 'all'
+  // which actually means 'any device_type' will be a match.
+  if (Iter == SyclDeviceTypeMap.end()) {
+    DeviceType = info::device_type::all;
+  } else {
+    DeviceType = Iter->second;
+    ColonPos = FilterString.find(":", Cursor);
+    if (ColonPos != std::string::npos) {
+      Cursor = ColonPos + 1;
+    } else {
+      Cursor = Cursor + Iter->first.size();
+    }
+  }
```

Comment:
Not a blocking comment from me: 
Thank you for adding comments here. 
You could also shorten this code almost twice and address the second part of comment from @keryell regarding code duplication and possible usage of lamda or funcs by having such additional fix:
```suggestion
  size_t ColonPos;
  auto findElement = [&](auto &Element) {
    size_t Found = FilterString.find(Element.first, Cursor);
    if (Found == std::string::npos)
      return false;
    Cursor = Found;
    return true;
  };
  auto selectElement = [&](auto It, auto Map, auto EltIfNotFound) {
    if (It == Map.end())
      return EltIfNotFound;
    ColonPos = FilterString.find(":", Cursor);
    if (ColonPos != std::string::npos)
      Cursor = ColonPos + 1;
    else
      Cursor = Cursor + It->first.size();
    return It->second;
  };
  // Check if the first entry matches with a known backend type
  auto It = std::find_if(
      std::begin(SyclBeMap), std::end(SyclBeMap), findElement);
  // If no match is found, set the backend type backend::all
  // which actually means 'any backend' will be a match.
  Backend = selectElement(It, SyclBeMap, backend::all);

  // Handle the optional 2nd field of the filter - device type.
  // Check if the 2nd entry matches with any known device type.
  auto Iter = std::find_if(
      std::begin(SyclDeviceTypeMap), std::end(SyclDeviceTypeMap), findElement);
  // If no match is found, set device_type 'all',
  // which actually means 'any device_type' will be a match.
  DeviceType = selectElement(Iter, SyclDeviceTypeMap, info::device_type::all);
```